crates/datasources/src/json/stream.rs

```diff
@@ -0,0 +1,83 @@
+use std::pin::Pin;
+use std::sync::{Arc, Mutex};
+use std::task::{Context, Poll};
+
+use datafusion::arrow::datatypes::{Schema, SchemaRef};
+use datafusion::arrow::json::ReaderBuilder;
+use datafusion::arrow::record_batch::RecordBatch;
+use datafusion::error::DataFusionError;
+use datafusion::execution::TaskContext;
+use datafusion::physical_plan::streaming::PartitionStream;
+use datafusion::physical_plan::{RecordBatchStream, SendableRecordBatchStream};
+use futures::{Stream, StreamExt};
+use serde_json::{Map, Value};
+
+type SendableCheckedRecordBatchStrem =
+    Pin<Box<dyn Stream<Item = Result<RecordBatch, DataFusionError>> + Send>>;
+
+pub struct JsonStream {
+    schema: Arc<Schema>,
+    stream: SendableCheckedRecordBatchStrem,
+}
+
+impl Stream for JsonStream {
+    type Item = Result<RecordBatch, DataFusionError>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        self.stream.poll_next_unpin(cx)
+    }
+}
+
+impl RecordBatchStream for JsonStream {
+    fn schema(&self) -> SchemaRef {
+        self.schema.clone()
+    }
+}
+
+pub struct JsonPartitionStream {
+    schema: Arc<Schema>,
+    stream: Mutex<Option<SendableCheckedRecordBatchStrem>>,
+}
+
+impl PartitionStream for JsonPartitionStream {
+    fn schema(&self) -> &SchemaRef {
+        &self.schema
+    }
+
+    fn execute(&self, _ctx: Arc<TaskContext>) -> SendableRecordBatchStream {
+        let partition = self
+            .stream
+            .lock()
+            .unwrap()
+            .take()
+            .expect("stream to only be called once")
+            .boxed();
+
+        Box::pin(JsonStream {
+            schema: self.schema.clone(),
+            stream: partition,
+        })
+    }
+}
+
+impl JsonPartitionStream {
+    pub fn new(schema: Arc<Schema>, chunk: Vec<Map<String, Value>>) -> Self {
+        let stream_schema = schema.clone();
+        let stream = futures::stream::iter(chunk)
+            .chunks(25)
```

Comment:
why 25?

Reply:
it's less than 100; Other numbers could be good too.

Reply:
Do you have a suggestion?

Reply:
I guess i don't understand why we chunk it, then chunk it again?

Reply:
these streaming tables work by collecting a number of streams (from files, though in this case we will have read them all into memory because we need/want to get the schema (and do that losslessly)). Then the stream produces record batches.

The first chunk, gives us the partitions, the second gives us the size of the record batch. 

I (following from your earlier comment) made the recordbatch size 1000, and the stream size as 10,000.